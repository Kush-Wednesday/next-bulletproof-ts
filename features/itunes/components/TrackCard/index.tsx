import React, { useRef, useState } from "react";
import { SongItem } from "../../api/getSongs";
import {
  CustomTrackCard,
  TrackName,
  ArtistName,
  AlbumArt,
  Duration,
  StyledAudio,
} from "@features/itunes/components/styled";
import { isUndefined } from "lodash";
import { injectIntl, IntlShape } from "react-intl";
import { T } from "../../../../common";
import { useRouter } from "next/router";

interface TrackCardProps {
  intl: IntlShape;
  result: SongItem;
  searchedTerm:string;
  memoizedAudioRef;
  trackEventHandler;
}

const TrackCard: React.FC<TrackCardProps> = props => {
  const { result, memoizedAudioRef, trackEventHandler,searchedTerm } = props;
  const BlockText = props => <T display="block" {...props} />;
  const router = useRouter();

  return (
    <CustomTrackCard data-testid="track-card">
      <AlbumArt
        data-testid="album-art"
        src={result.artworkUrl100}
        onClick={() => router.push(`/trackDetails/${result.trackId}?song=${searchedTerm}`)}
      ></AlbumArt>
      <TrackName>
        <BlockText
          id="track_name"
          values={{ name: !isUndefined(result.trackName) && result.trackName.substring(0, 10) }}
        />
      </TrackName>
      <ArtistName>
        <BlockText
          id="artist_name"
          values={{ name: !isUndefined(result.trackName) && result.artistName.substring(0, 10) }}
        />
      </ArtistName>
      <BlockText id="duration" />
      {result.trackTimeMillis && Math.floor(result.trackTimeMillis / 60000)}:
      {result.trackTimeMillis && Math.floor((result.trackTimeMillis / 1000) % 60)}s
      <StyledAudio>
        <audio
          id={result.previewUrl}
          data-testid="audio-test"
          aria-label="audio-label"
          autoPlay={false}
          controls
          onPlay={() => trackEventHandler(result.trackId)}
          ref={memoizedAudioRef ? memoizedAudioRef?.[result.trackId] : null}
          style={
            {
              display: "block",
              width: "fit-content<100%>",
              maxWidth: "100%",
              color: "c9ada7",
            } as React.CSSProperties
          }
        >
          <source src={result.previewUrl} type="audio/mp3"></source>
          Your browser does not support audio tags
        </audio>
      </StyledAudio>
    </CustomTrackCard>
  );
};

export default injectIntl(TrackCard);
