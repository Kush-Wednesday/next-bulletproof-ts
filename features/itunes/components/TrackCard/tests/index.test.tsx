import { fireEvent, render } from "@utils/testUtils";
import TrackCard from "../index";

describe("<TrackCard />", () => {
  const trackCardProps = {
    result: {
      artistName: "Adele",
      trackName: "Hello",
      trackId: 1544494392,
      artistId: 262836961,
      trackTimeMillis: 295502,
      artworkUrl100:
        "https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg",
      previewUrl: "22233223",
    },
    memoizedAudioRef: {},
  };
  let trackEventSpy;
  beforeEach(() => {
    trackEventSpy = jest.fn();
  });

  it("should match the snapshot", () => {
    const { baseElement } = render(
      <TrackCard {...trackCardProps} trackEventHandler={trackEventSpy} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it(" display dom for debug", () => {
    const { debug } = render(<TrackCard {...trackCardProps} trackEventHandler={trackEventSpy} />);
    debug();
  });

  it("should check if image is rendered", () => {
    const { getAllByTestId } = render(
      <TrackCard {...trackCardProps} trackEventHandler={trackEventSpy} />
    );
    const albumArt = getAllByTestId("album-art")[0] as HTMLImageElement;
    expect(albumArt.src).toContain(
      "https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg"
    );
  });

  it("should render audio element", async () => {
    const { getAllByTestId, debug } = await render(
      <TrackCard {...trackCardProps} trackEventHandler={trackEventSpy} />
    );
    const audioElement = getAllByTestId("audio-test")[0] as HTMLAudioElement;
    debug(audioElement);
    fireEvent.play(audioElement);
    expect(trackEventSpy).toBeCalled();
    expect(audioElement.id).toContain("22233223");
    expect(audioElement.controls).toBeTruthy();
  });
});
