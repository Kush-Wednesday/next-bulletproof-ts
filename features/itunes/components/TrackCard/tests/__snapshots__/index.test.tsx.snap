// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered styled__CustomCard-sc-15gkald-1 styled__CustomTrackCard-sc-14jlyoy-2 kmxBSK liLTAq"
      data-testid="track-card"
    >
      <div
        class="ant-card-body"
      >
        <img
          class="styled__AlbumArt-sc-14jlyoy-5 ivmedU"
          data-testid="album-art"
          src="https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg"
        />
        <h3
          class="styled__TrackName-sc-14jlyoy-3 lowavE"
        >
          <p
            class="T__StyledText-sc-apdcd0-0 fjJkvM"
            data-testid="t"
            display="block"
          >
            Track Name: Hello
          </p>
        </h3>
        <h3
          class="styled__ArtistName-sc-14jlyoy-4 iUIEPX"
        >
          <p
            class="T__StyledText-sc-apdcd0-0 fjJkvM"
            data-testid="t"
            display="block"
          >
            Artist Name: Adele
          </p>
        </h3>
        <p
          class="styled__Duration-sc-14jlyoy-6 jcLzMV"
        >
          <p
            class="T__StyledText-sc-apdcd0-0 fjJkvM"
            data-testid="t"
            display="block"
          >
            Duration :
          </p>
          4
          :
          55
          s
        </p>
        <div
          class="styled__StyledAudio-sc-14jlyoy-7 ggvosY"
        >
          <audio
            aria-label="audio-label"
            controls=""
            data-testid="audio-test"
            id="22233223"
            style="display: block; max-width: 100%;"
          >
            <source
              src="22233223"
              type="audio/mp3"
            />
            Your browser does not support audio tags
          </audio>
        </div>
      </div>
    </div>
  </div>
</body>
`;
